### PR TITLE
fix(markdown): respect CommonMark backtick delimiters during currency escaping

### PR DESCRIPTION
**Pull request by uczltw6**
`escapeCurrencyDollars()` skipped code by finding the next occurrence of the opening backtick string. a longer run inside an inline code span, or a same-length run inside a fenced block, could be mistaken for the closer, so leftover `$5` was rewritten to `\$5`.

the scanner now uses CommonMark rules: inline spans close on an equal-length run; fenced blocks close only on a standalone fence line at the same blockquote depth as the opener (list markers are not valid closers). an unclosed fence owns the rest of the input, which is the streaming case.

mirrored in `@assistant-ui/react-markdown` and `@assistant-ui/react-streamdown`.

<a href="https://rupic.dev/s/s2.P7HYUct0DvEPTTSAgZRUsqL4OyE80BzmrH13EjDMI73wciLr0lXswOiirB8?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### CODE REVIEW

File: packages/react-markdown/src/preprocess.ts [hunk 1]
```diff
@@ -72,9 +72,40 @@ function runLength(text: string, start: number, char: string): number {
  * text.
  */
 function codeSpanEnd(text: string, start: number): number {
-  const fence = "`".repeat(runLength(text, start, "`"));
-  const closed = text.indexOf(fence, start + fence.length);
-  return closed === -1 ? -1 : closed + fence.length;
+  const delimiterLength = runLength(text, start, "`");
+  const lineStart = text.lastIndexOf("\n", start - 1) + 1;
+  const openingIndent = text.slice(lineStart, start);
+  const openingLineEnd = text.indexOf("\n", start + delimiterLength);
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts]: Fixed in 9507428. Line-boundary helpers now handle CR, LF, and CRLF for both opening and closing fence scans; the mirrored CR-only regression passes.

File: packages/react-streamdown/src/preprocess.ts [hunk 2]
```diff
@@ -72,9 +72,40 @@ function runLength(text: string, start: number, char: string): number {
...
+
+  if (isFence) {
+    if (openingLineEnd === -1) return -1;
+    const closingFence = new RegExp(
+      "^ {0,3}(`{" + delimiterLength + ",})[\\t ]*\\r?$",
+      "gm",
+    );
+    closingFence.lastIndex = openingLineEnd + 1;
+    const match = closingFence.exec(text);
+    if (!match) return -1;
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-streamdown/src/preprocess.ts]: Fixed in 9507428. An unterminated recognized fence now preserves the remainder of the input through EOF, with mirrored regression coverage in both markdown packages.

File: packages/react-markdown/src/preprocess.ts [hunk 3]
```diff
@@ -72,9 +72,40 @@ function runLength(text: string, start: number, char: string): number {
...
+  const openingInfo = text.slice(
+    start + delimiterLength,
+    openingLineEnd === -1 ? text.length : openingLineEnd,
+  );
+  const isFence =
+    delimiterLength >= 3 &&
+    /^ {0,3}$/.test(openingIndent) &&
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts]: Fixed in 9507428. Fence-prefix parsing now recognizes blockquote and list-item container prefixes before applying the fence rules, with mirrored blockquote and list regressions.

File: packages/react-markdown/src/preprocess.ts [hunk 4]
```diff
@@ -72,9 +72,40 @@ function runLength(text: string, start: number, char: string): number {
...
+
+  if (isFence) {
+    if (openingLineEnd === -1) return -1;
+    const closingFence = new RegExp(
+      "^ {0,3}(`{" + delimiterLength + ",})[\\t ]*\\r?$",
+      "gm",
+    );
+    closingFence.lastIndex = openingLineEnd + 1;
+    const match = closingFence.exec(text);
+    if (!match) return -1;
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts]: Fixed in 9507428. Recognized fenced blocks now extend to `text.length` when no valid closing fence exists, while `-1` remains limited to unclosed inline spans. Mirrored unclosed-fence regressions pass in both packages, and the function documentation now distinguishes the two cases.

File: packages/react-streamdown/src/preprocess.ts [hunk 5]
```diff
@@ -66,15 +66,125 @@ function runLength(text: string, start: number, char: string): number {
...
+    const blockquote = rest.match(/^ {0,3}>[ \t]?/);
+    if (!blockquote) break;
+    rest = rest.slice(blockquote[0].length);
+    depth += 1;
+  }
+  return { depth, rest };
+}
+
+function isFencePrefix(prefix: string): boolean {
+  let remaining = peelBlockquotes(prefix).rest;
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-streamdown/src/preprocess.ts]: Fixed in a302a81. Fence prefixes are now parsed as an ordered sequence of list and blockquote containers, and closing/content lines consume that same sequence. Mirrored `- > ```...` regressions pass in both packages.

File: packages/react-markdown/src/preprocess.ts [hunk 6]
```diff
@@ -66,15 +66,125 @@ function runLength(text: string, start: number, char: string): number {
...
+    const listItem = remaining.match(/^ {0,3}(?:[-+*]|\d{1,9}[.)])[ \t]+/);
+    if (!listItem) break;
+    remaining = remaining.slice(listItem[0].length);
+  }
+  return /^ {0,3}$/.test(remaining);
+}
+
+function isClosingFencePrefix(prefix: string, quoteDepth: number): boolean {
+  const peeled = peelBlockquotes(prefix);
+  return peeled.depth === quoteDepth && /^ {0,3}$/.test(peeled.rest);
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts]: Fixed in a302a81 and covered further in 1b3b825. The opening context now records each list container's content indentation, so closers are measured relative to that column. Mirrored regressions cover a nested-list content column and a `10. ` ordered marker.

File: packages/react-markdown/src/preprocess.ts [hunk 7]
```diff
@@ -66,15 +66,125 @@ function runLength(text: string, start: number, char: string): number {
...
+        closingLineStart,
+        closingLineEnd,
+        delimiterLength,
+        peelBlockquotes(openingPrefix).depth,
+      );
+      if (close !== -1) return close;
+      if (closingLineEnd === text.length) break;
+      closingLineStart = nextLineStart(text, closingLineEnd);
+    }
+    return text.length;
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts:292]: Fixed in a302a81. The scan now checks each content line against the opening container sequence and returns at the container boundary instead of protecting the rest of the document. The blockquote example now escapes the outside `$5`.

File: packages/react-markdown/src/preprocess.test.ts [hunk 8]
```diff
@@ -132,6 +132,72 @@ describe("escapeCurrencyDollars", () => {
...
+    ).toBe("```\nconst marker = ```;\nconst price = $5;\n```\nafter \\$7");
+  });
+
+  it("accepts a closing fence longer than the opening fence", () => {
+    expect(escapeCurrencyDollars("```\n$5\n````\nafter `code $7` and $9")).toBe(
+      "```\n$5\n````\nafter `code $7` and \\$9",
+    );
+  });
+
+  it("does not rewrite currency inside an unclosed fenced block", () => {
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.test.ts:155]: Added in a302a81. Both package suites now include mirrored unclosed-fence regressions for blockquote and list containers; the full suites pass with 53 and 147 tests.

**Review by okisdev — CHANGES_REQUESTED** [review]
request changes. this PR is still growing while review and CI are in flight (two more commits landed after the closer-prefix fix and restarted the wait set). if the scanner is not finished, convert the PR to a draft and mark it ready only when the implementation and tests are done. a ready PR that keeps moving mid-review cannot be approved.

**Comment by uczltw6** [comment]
Thanks — I converted the PR to draft before this final push. The outstanding tab-expanded list-indentation case is fixed in 59ccd5f with mirrored regressions for both packages. The complete package suites now pass (54 react-markdown tests and 148 react-streamdown tests), along with both type checks, Oxlint, Oxfmt, and git diff checks. I will wait for the refreshed CI and review runs to settle before marking it ready again.

File: packages/react-markdown/src/preprocess.test.ts [hunk 9]
```diff
@@ -132,6 +132,110 @@ describe("escapeCurrencyDollars", () => {
...
+    expect(
+      escapeCurrencyDollars(
+        "- > ```\n  > const marker = ```;\n  > const price = $5;\n  > ```\nafter $7",
+      ),
+    ).toBe(
+      "- > ```\n  > const marker = ```;\n  > const price = $5;\n  > ```\nafter \\$7",
+    );
+  });
+
+  it("recognizes fenced blocks at a nested list content column", () => {
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.test.ts:199]: Added in 59ccd5f. Both mirrored suites cover a tab-separated list opener, tab-indented body and closer, plus a container-boundary case that distinguishes two outside spaces from the four-column list content indent. Full suites pass with 54 and 148 tests.

File: packages/react-markdown/src/preprocess.ts [hunk 10]
```diff
@@ -66,15 +66,181 @@ function runLength(text: string, start: number, char: string): number {
...
+    const blockquote = remaining.match(/^ {0,3}>[ \t]?/);
+    if (blockquote) {
+      remaining = remaining.slice(blockquote[0].length);
+      containers.push({ type: "blockquote" });
+      continue;
+    }
+    const listItem = remaining.match(/^ {0,3}(?:[-+*]|\d{1,9}[.)])[ \t]+/);
+    if (listItem) {
+      remaining = remaining.slice(listItem[0].length);
+      containers.push({ type: "list", indent: listItem[0].length });
```
[the remark this answers is not in the document]
Comment (uczltw6) [reply packages/react-markdown/src/preprocess.ts]: Fixed in 59ccd5f. List-container indentation is now stored and consumed in tab-expanded visual columns, including the current opening column after preceding containers. The exact tab-separated list regression now preserves the fenced `$5` and escapes the outside `$7` in both packages.

**Comment by okisdev** [comment]
closing. #6160 fixes the reproducible inline equal-length delimiter bug without expanding the currency heuristic into a CommonMark fence and container parser. the fence-in-fence behavior this PR grew to handle remains the accepted #5160 limitation.
